packages/flutter/lib/src/semantics/semantics.dart

```diff
@@ -4014,17 +4002,35 @@ abstract class SemanticsSortKey extends Diagnosticable implements Comparable<Sem
   /// const constructors so that they can be used in const expressions.
   const SemanticsSortKey({this.name});
 
-  /// An optional name that will make this sort key only order itself
-  /// with respect to other sort keys of the same [name], as long as
-  /// they are of the same [runtimeType].
+  /// An optional name that will group this sort key with other sort keys of the
+  /// same [name].
+  ///
+  /// Sort keys must have the same `runtimeType` when compared.
+  ///
+  /// Keys with no [name] are compared to to other keys with no [name], and will
+  /// be traversed before those with a [name].
   final String name;
 
   @override
   int compareTo(SemanticsSortKey other) {
-    // The sorting algorithm must not compare incomparable keys.
-    assert(runtimeType == other.runtimeType);
-    assert(name == other.name);
-    return doCompare(other);
+    // Sort by name first and then subclass ordering.
+    assert(runtimeType == other.runtimeType, 'Semantics sort keys can only be compared to other sort keys of the same type.');
+
```

Comment:
If they aren't the same `runtimeType`, should we return here for non-debug builds?

Reply:
One of the main reasons to use asserts to enforce the contract is to avoid a lot of code for checking the contract in release mode that slows things down, and this, while it's just a conditional, just adds overhead to programs that abide by the contract.  Once the contract is violated, then all bets are off anyhow.

Also, what should we return? It seems like it's possible that it would do the "right" thing if they have different types, but at least one of them has a name.